### DISCUSSION TITLE
Higher Parallelism Memory Usage and Merging Parallel Results

### DISCUSSION THREAD

**Discussion by Korg95** [Beuth-Erdelt/DBMS-Benchmarker#173] [General]
Context

Welcome and nice to meet you!

Thanks for the kind words. Could you tell us something about what your use case is? Which database system are you interested in? What is the workload? Are there any options or parameters that you particularly use (or miss)?

Thank you and welcome again!

========================================================================

Hi @perdelt ,

Apologies for the late reply!

Currently I am using the tool to test online analytical workloads across 3 databases. Vertica, Druid and Clickhouse.
It's a bit interesting compared to what classical workloads go through this tool in the sense that I have to test for low latencies < 1-2 seconds and around 60-100 QPS with randomized query execution to simulate daily users executing a variety of tasks in a concurrent manner. This I hope showcases the effect of noisy neighbors if any.

So far comparing the databases is a breeze when working with lower amounts of threads, I however run into issues when running 60 threads or more with JVM OOM issues in relation to the variety of JDBC drivers that using. It hopefully should be solvable and I can reach back with more info on that side once I get more time.

What I've found more interesting is figuring out how to simplify the merging of results. When testing with parallelism it results in many subfolders being generated with results that have connections renamed according to numbered suffixes assigned. I can't seem to merge those results into a single result. This means that when comparing results each of these connections is considered a different DBMS. It is fairly easy myself to merge these adhoc but I was wondering if this is a case of user error or some misunderstanding of how the tooling works.

Here is my workflow:

python .\DBMS-Benchmarker\benchmark.py run -b -d -w connection -f ./config -r ./results -p 10 -pp -ssh 1
python .\DBMS-Benchmarker\benchmark.py read -r ./results/1721340615 -e yes
python .\DBMS-Benchmarker\merge.py -r ./results/ -c 1721340615 #Does not yield merging of results as expected
Note here I rely on -ssh query randomization to give me some sense of mixing up what queries run concurrently with each other. Since the tool processes randomized query by query I rely on 10 processes to ensure that different workloads execute at the same time, collecting a large enough sample size of runs per query.

I also set a query cool down to help rate limit the queries executed so that I can get closer to 60 QPS. 60 QPS / 10 threads = 6 QPS per thread = 1 second / 6 QPS = 0.166 recurring cooldown per query. This logic might be flawed though, where I should only apply a cooldown if I am getting too many failures and not finding the QPH that your tool outputs.

Small note: I think that for this case, QPS is more appropriate than QPH but that's a very small note :)

(PS if you want me to start a discussion let me know and I can move it there)

Kind regards,

Korg95

========================================================================

Hi @Korg95 ,

Thank you for sharing your experiences!

I see that you have become familiar with the tool well.

As short quick answers:

Yes, please feel free to open an issue if you want to discuss specific points.

Have you tried this?

python .\DBMS-Benchmarker\benchmark.py run -b -d -w connection -f ./config -r ./results -p 10 -pp -ssh 1 -e yes
So integrating the generation of the evaluation into the execution directly? This should also do the merging of results.
I will test why the single calls in three parts do not work.

PS: You might want to try the silent mode -vn. This suppresses most of the output and can be useful if you have a lot of fast queries (it saves a bit of overhead).

========================================================================

I'll give it a spin and report back in a discussion thread moving the context there. Cheers @perdelt

**Comment by Korg95** [comment]
So I tried @perdelt I tried
`python .\DBMS-Benchmarker\benchmark.py run -b -d -w connection -f ./config -r ./results -p 2 -pp -ssh 1 -e yes -vn`
Which yields
![image](https://github.com/user-attachments/assets/c91f4fd7-5f90-48f1-8924-c4f09042f365)

The output does mention the merge step but we don't seem to get the merge behaviour.
![image](https://github.com/user-attachments/assets/a1d34da0-39e5-4329-8923-f7860662864d)

If I look at the end the QpH does actually indicate a merged results which is good. At the query level it doesn't seem to do it.

**Comment by perdelt** [comment]
Hi @Korg95 ,

I see. At the moment, the tool only shows query statistics per stream, not aggregated across parallel streams. You can create your own evaluations using the Python interface.

The suggestion seems sensible to me, though. (However, the effect on individual queries can depend on chance if you randomly choose the order of execution.)

Please try the following and give me feedback. If it does what you expect, I'll clean up the code and integrate it into the next release.

```python
from dbmsbenchmarker import *
import pandas as pd

resultfolder = "./results/"
code = "1721684817"

evaluate = inspector.inspector(resultfolder)
evaluate.load_experiment(code)

list_connections = evaluate.get_experiment_list_connections()
list_queries = evaluate.get_experiment_list_queries()
query_properties = evaluate.get_experiment_query_properties()
def map_index_to_queryname(numQuery):
    if numQuery in query_properties and 'config' in query_properties[numQuery] and 'title' in query_properties[numQuery]['config']:
        return query_properties[numQuery]['config']['title']
    else:
        return numQuery

dbms_list = {}
for connection_nr, connection in evaluate.benchmarks.dbms.items():
    c = connection.connectiondata
    if 'orig_name' in c:
        orig_name = c['orig_name']
    else:
        orig_name = c['name']
    if not orig_name in dbms_list:
        dbms_list[orig_name] = [c['name']]
    else:
        dbms_list[orig_name].append(c['name'])

for numQuery in list_queries:
    print("# Query Q{}: {}".format(numQuery, map_index_to_queryname(str(numQuery))))
    df1, df2 = evaluate.get_measures_and_statistics(numQuery, type='timer', name='execution')
    reverse_mapping = {sub_db: main_db for main_db, sub_dbs in dbms_list.items() for sub_db in sub_dbs}
    df1['DBMS'] = df1.index.map(reverse_mapping)
    df_melted = pd.melt(df1, id_vars=['DBMS'], value_vars=df1.columns).sort_values('DBMS')
    df_melted = pd.DataFrame(df_melted[['DBMS', 'value']])
    df_results = pd.DataFrame()
    df_groups = df_melted.groupby('DBMS')
    group_list = list(df_groups.groups.keys())
    for group in df_groups:
        df_dbms = pd.DataFrame(group[1]['value']).T
        df_dbms.index = [group[0]]
        df_dbms = evaluator.addStatistics(df_dbms, drop_measures=True)
        df_results = pd.concat([df_results, df_dbms], ignore_index=False)
    print(df_results.round(2))
```

**Reply by Korg95** [reply]
Thanks for your quick response against this. I do agree that the random selection might end up making things not work as expected.

I think that the closest thing to this is that ability to have a stream select a query list ordering to run? I need to check that out in the options after I try this and report back on how it works. If not that I could make a change to pre-calculate the randomized orderings and then assign by thread dynamically via query config. Sounds like more work though I'll read into the options and report back.

Cheers,

Korg95

**Reply by perdelt** [reply]
Have you tried the code snippet? Does it do what you want?

Maybe you can use it to judge whether the effect is as expected.

**Reply by Korg95** [reply]
Sorry, been a busy day. It worked as expected! I got all my results merged correctly by query using your methodology.

I'll start running some wider tests to fully validate this and move onto tackling parallelism OOMs.

**Comment by perdelt** [comment]
Hi @Korg95 

I appreciate your feedback!

There is a new version v0.14.3 you may want to try. I improved the output a bit for parallel streams and shuffled query ordering.

Moreover you now may submit options to the JVM:
```
    {
        'name': 'PostgreSQL',
        'info': 'This is a demo of PostgreSQL',
        'active': True,
        'dialect': 'PostgreSQL',
        'JDBC': {
            'driver': 'org.postgresql.Driver',
            'url': 'jdbc:postgresql://localhost:5432/tpch',
            'auth': ['postgres', 'postgres'],
            'jar': 'jars/postgresql-42.2.5.jar',
            'options': ["-Xms312m", "-Xmx312m"],
        }
    },
```

Maybe this will help with your OOM problem (but I assume you might need more RAM).

Maybe you also want to try

* `-s 1234` sets a random seed (to 1234). This should improve reproducibility and also affects random query ordering.
* `-pn 4` repeats every query (with different parameters) 4 times in a row. This should give more stability and improve "per query" evaluation.

If you want to set a fixed order of queries per stream, you can find an example of `stream_ordering` in example/tpc-h/queries.config.

**Reply by Korg95** [reply]
Hi @perdelt,

I spent about a day or so attempting to optimize memory usage. I was able to do exactly what you were showing by setting global JVM options that were indeed getting picked up. I actually opted for 512m to see if that would be enough but kept running into OOM. This is on a box with 128GB of ram.

I decided to try in the final target language of C# doing the same test and also observed connection establishment timing reduction, latency differences between the python call to jdbc and I was able to ramp up 100s of connections with minimal memory footprint using a ado.net driver for Clickhouse.

Unfortunately for my high throughput use case and the time I have left to perform the evaluation I need to pivot to that approach and the final decision for me is that this kind of use case has a memory leak reported in the JDBC library itself https://github.com/baztian/jaydebeapi/issues/227

Appreciate all your time on this, the use of your program and stats involved has been a good experience otherwise!

**Reply by perdelt** [reply]
Hi @Korg95 

Oh, I didn't know that. Maybe one has to call a GC every now and then or something. Yes, then it makes sense to change the client.

Thank you for the link and your feedback!


Best regards!

**Reply by Korg95** [reply]
Pleasure communicating with you @perdelt! Take care :)
